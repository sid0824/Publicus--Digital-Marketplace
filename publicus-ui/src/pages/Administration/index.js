import React from 'react';
import {Button, Collapse, Input, message} from 'antd';
import {API, Auth} from "aws-amplify";
import {ApiName} from "../../core/Config";
import {Link} from "react-router-dom";
import RoutePaths from "../../core/RoutePaths";
import FormImage from "../../components/FormImage";
import {FormPublishedStates} from "../MyForms";

const { Panel } = Collapse;

/**
 * Page for administrators to approve and reject forms along with other admin actions
 */
export default class Administration extends React.Component {

  state = {
    approving: null, // Currently approving
    rejecting: null, // Currently rejecting
    newAdminEmail: '', // Email of new admin
    loadingNewAdmin: false, // Saving the data of a new admin
    toApprove: [] // List of forms to approve
  };

  componentDidMount = async () => {
    await this.fetchData();
    const authData = await Auth.currentCredentials();
    const { IdentityId } = authData.params;
    this.setState({
      IdentityId
    })
  };

  /**
   *
   * @returns {Promise<void>}
   */
  fetchData = async () => {
    try {
      const toApprove = await API.get(ApiName, "/viewFormsForApproval", {});
      this.setState({
        toApprove
      })
    } catch (err) {
      console.error(err);
    }
  };

  /**
   * Approve a form for the marketplace
   * @param form - form to approve
   */
  approve = (form) => async (e) => {
    e.stopPropagation();
    this.setState({
      approving: form.surveyId
    });
    try {
      await API.post(ApiName, "/modifySurveyStatus", {
        body: {
          status: FormPublishedStates.PUBLISHED,
          surveyId: form.surveyId
        }
      });
      message.success('Approved form');
      await this.fetchData();
    } catch (err) {
      console.error(err);
      message.error("Failed to approve form")
    }
    this.setState({
      approving: null
    })
  };

  /**
   * Reject a form from the marketplace
   * @param form - form to reject
   */
  reject = (form) => async (e) => {
    e.stopPropagation();
    this.setState({
      rejecting: form.surveyId
    });
    try {
      await API.post(ApiName, "/modifySurveyStatus", {
        body: {
          status: FormPublishedStates.REJECTED,
          surveyId: form.surveyId
        }
      });
      message.success('Rejected form');
      this.fetchData();
    } catch (err) {
      console.error(err);
      message.error("Failed to reject form")
    }
    this.setState({
      rejecting: false
    })
  };

  /**
   * Create a list header for a form
   * @param form
   */
  getHeader = (form) => {
    const { approving, rejecting } = this.state;
    return (
      <div style={{ display: 'flex', flexDirection: 'row', justifyContent: 'space-between', minWidth: 170 }}>
        <span style={{ marginTop: 4 }}>{unescape(form.name)}</span>
        <div>
          <Button
            type="primary"
            onClick={this.approve(form)}
            loading={approving === form.surveyId}
          >
            Approve
          </Button>
          <Button
            type="danger"
            onClick={this.reject(form)}
            style={{ marginLeft: 4 }}
            loading={rejecting === form.surveyId}
          >
            Reject
          </Button>
        </div>
      </div>
    )
  };

  /**
   * Turn a user into an administrator
   */
  onAddAdmin = async () => {
    const { newAdminEmail } = this.state;
    this.setState({
      loadingNewAdmin: true
    });
    try {
      await API.post(ApiName, "/createAdmin", {
        body: {
          email: newAdminEmail
        }
      });
      message.success(`Successfully made ${newAdminEmail} into an administrator`);
    } catch (err) {
      console.error(err);
      message.error(`Failed to make ${newAdminEmail} into an administrator`)
    }
    this.setState({
      loadingNewAdmin: false
    });
  };

  /**
   * Update the admin email
   * @param e - change event
   */
  changeAdminEmail = (e) => {
    this.setState({
      newAdminEmail: e.target.value
    })
  };

  /**
   * Render the content of a form list
   */
  renderToApprove = () => {
    const { toApprove, IdentityId } = this.state;
    const panels = toApprove.map(survey => {
      return (
        <Panel header={this.getHeader(survey)} key={survey.surveyId}>
          <Link
            to={`${RoutePaths.CompleteForm}?surveyId=${survey.surveyId}&creatorId=${(IdentityId)}`}
          >
            View Form
          </Link>
          <br />
          <br />
          <Link
            to={`${RoutePaths.Marketplace}?search=${survey.userId}`}
          >
            View Other Forms By This User
          </Link>
          <br />
          <br />
          <strong>Form Image</strong>
          <br />
          <FormImage
            image={survey.image}
          />
        </Panel>
      )
    });
    return(
      <React.Fragment>
        {panels}
      </React.Fragment>
    )
  };

  render() {
    const { newAdminEmail, loadingNewAdmin, toApprove } = this.state;
    return(
      <div style={{ display: 'flex', flexDirection: 'row', height: '100%', marginLeft: '12%', marginRight: '12%', marginTop: 10 }}>
        <div style={{ width: '50%', marginRight: 15 }}>
          <h2>Administrator Actions</h2>
          <span>Make New Administrator</span>
          <Input
            placeholder={"Enter the email address of a new administrator"}
            value={newAdminEmail}
            onChange={this.changeAdminEmail}
            style={{ marginBottom: 3, width: 'calc(100% - 120px)'}}
          />
          <Button
            type={"primary"}
            loading={loadingNewAdmin}
            onClick={this.onAddAdmin}
          >
            Submit
          </Button>
        </div>
        <div style={{ width: '50%'}}>
          <h2>Forms To Review</h2>
          <div style={{ height: 'calc(100% - 50px)', overflowY: 'scroll' }}>
            {
              toApprove.length
                ? <Collapse accordion>
                    {this.renderToApprove()}
                  </Collapse>
                : <strong>No forms currently require approval</strong>
            }
          </div>
        </div>
      </div>
    )
  }
}