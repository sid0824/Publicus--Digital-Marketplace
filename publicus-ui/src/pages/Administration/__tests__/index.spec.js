import toJson from 'enzyme-to-json';
import React from "react";
import {shallow, configure} from 'enzyme';
import Forms from '../../../fixtures/Forms';
import Adapter from 'enzyme-adapter-react-16';
import Administration from "../index";
import {API, Auth} from "aws-amplify";
import { message } from 'antd';

configure({ adapter: new Adapter() });

jest.mock('antd');

describe('Administration', () => {

  beforeEach(() => {
    Auth.currentCredentials = jest.fn().mockResolvedValue({
      params: {
        IdentityId: 'MOCK_IDENTITY'
      }
    });
    API.get = jest.fn().mockResolvedValue(Forms);
    API.post = jest.fn();
  });

  afterEach(() => {
    Auth.currentCredentials.mockReset();
    API.get.mockReset();
    API.post.mockReset();
    message.success.mockClear();
  });

  test('should render administration page', (done) => {
    const wrapper = shallow(
      <Administration />
    );
    setTimeout(() => {
      expect(toJson(wrapper)).toMatchSnapshot();
      done();
    })
  });

  test('should approve form', async () => {
    const wrapper = shallow(
      <Administration />
    );
    await wrapper.instance().approve(Forms[0])({ stopPropagation: () => {}});
    expect(API.post.mock.calls[0]).toMatchSnapshot();
    expect(API.get.mock.calls[1]).toMatchSnapshot();
    expect(message.success.mock.calls[0]).toMatchSnapshot();
  });

  test('should reject form', async () => {
    const wrapper = shallow(
      <Administration />
    );
    await wrapper.instance().reject(Forms[0])({ stopPropagation: () => {}});
    expect(API.post.mock.calls[0]).toMatchSnapshot();
    expect(API.get.mock.calls[1]).toMatchSnapshot();
    expect(message.success.mock.calls[0]).toMatchSnapshot();
  });

  test('should add admin', async () => {
    const wrapper = shallow(
      <Administration />
    );
    wrapper.setState({
      newAdminEmail: 'MOCK_EMAIL'
    });
    await wrapper.instance().onAddAdmin();
    expect(API.post.mock.calls[0]).toMatchSnapshot();
  });

});