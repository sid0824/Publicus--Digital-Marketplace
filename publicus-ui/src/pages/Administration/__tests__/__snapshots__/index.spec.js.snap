// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Administration should add admin 1`] = `
Array [
  "prod-publicus-app-api",
  "/createAdmin",
  Object {
    "body": Object {
      "email": "MOCK_EMAIL",
    },
  },
]
`;

exports[`Administration should approve form 1`] = `
Array [
  "prod-publicus-app-api",
  "/modifySurveyStatus",
  Object {
    "body": Object {
      "status": 3,
      "surveyId": "2d239420-f49d-11e9-8487-c5a7ea76fe8d",
    },
  },
]
`;

exports[`Administration should approve form 2`] = `
Array [
  "prod-publicus-app-api",
  "/viewFormsForApproval",
  Object {},
]
`;

exports[`Administration should approve form 3`] = `
Array [
  "Approved form",
]
`;

exports[`Administration should reject form 1`] = `
Array [
  "prod-publicus-app-api",
  "/modifySurveyStatus",
  Object {
    "body": Object {
      "status": 2,
      "surveyId": "2d239420-f49d-11e9-8487-c5a7ea76fe8d",
    },
  },
]
`;

exports[`Administration should reject form 2`] = `
Array [
  "prod-publicus-app-api",
  "/viewFormsForApproval",
  Object {},
]
`;

exports[`Administration should reject form 3`] = `
Array [
  "Rejected form",
]
`;

exports[`Administration should render administration page 1`] = `
<div
  style={
    Object {
      "display": "flex",
      "flexDirection": "row",
      "height": "100%",
      "marginLeft": "12%",
      "marginRight": "12%",
      "marginTop": 10,
    }
  }
>
  <div
    style={
      Object {
        "marginRight": 15,
        "width": "50%",
      }
    }
  >
    <h2>
      Administrator Actions
    </h2>
    <span>
      Make New Administrator
    </span>
    <Input
      onChange={[Function]}
      placeholder="Enter the email address of a new administrator"
      style={
        Object {
          "marginBottom": 3,
          "width": "calc(100% - 120px)",
        }
      }
      type="text"
      value=""
    />
    <Button
      block={false}
      ghost={false}
      htmlType="button"
      loading={false}
      onClick={[Function]}
      type="primary"
    >
      Submit
    </Button>
  </div>
  <div
    style={
      Object {
        "width": "50%",
      }
    }
  >
    <h2>
      Forms To Review
    </h2>
    <div
      style={
        Object {
          "height": "calc(100% - 50px)",
          "overflowY": "scroll",
        }
      }
    >
      <Collapse
        accordion={true}
        bordered={true}
        expandIconPosition="left"
        openAnimation={
          Object {
            "appear": [MockFunction],
            "enter": [MockFunction],
            "leave": [MockFunction],
          }
        }
      >
        <CollapsePanel
          header={
            <div
              style={
                Object {
                  "display": "flex",
                  "flexDirection": "row",
                  "justifyContent": "space-between",
                  "minWidth": 170,
                }
              }
            >
              <span
                style={
                  Object {
                    "marginTop": 4,
                  }
                }
              >
                a
              </span>
              <div>
                <Button
                  block={false}
                  ghost={false}
                  htmlType="button"
                  loading={false}
                  onClick={[Function]}
                  type="primary"
                >
                  Approve
                </Button>
                <Button
                  block={false}
                  ghost={false}
                  htmlType="button"
                  loading={false}
                  onClick={[Function]}
                  style={
                    Object {
                      "marginLeft": 4,
                    }
                  }
                  type="danger"
                >
                  Reject
                </Button>
              </div>
            </div>
          }
          key="2d239420-f49d-11e9-8487-c5a7ea76fe8d"
        >
          <Link
            to="/completeForm?surveyId=2d239420-f49d-11e9-8487-c5a7ea76fe8d&creatorId=MOCK_IDENTITY"
          >
            View Form
          </Link>
          <br />
          <br />
          <Link
            to="/marketplace?search=ap-southeast-2:91ecfa27-a7c0-45a8-9f45-c93885409ca7"
          >
            View Other Forms By This User
          </Link>
          <br />
          <br />
          <strong>
            Form Image
          </strong>
          <br />
          <FormImage
            image="https://designshack.net/wp-content/uploads/placeholder-image.png"
          />
        </CollapsePanel>
        <CollapsePanel
          header={
            <div
              style={
                Object {
                  "display": "flex",
                  "flexDirection": "row",
                  "justifyContent": "space-between",
                  "minWidth": 170,
                }
              }
            >
              <span
                style={
                  Object {
                    "marginTop": 4,
                  }
                }
              >
                a
              </span>
              <div>
                <Button
                  block={false}
                  ghost={false}
                  htmlType="button"
                  loading={false}
                  onClick={[Function]}
                  type="primary"
                >
                  Approve
                </Button>
                <Button
                  block={false}
                  ghost={false}
                  htmlType="button"
                  loading={false}
                  onClick={[Function]}
                  style={
                    Object {
                      "marginLeft": 4,
                    }
                  }
                  type="danger"
                >
                  Reject
                </Button>
              </div>
            </div>
          }
          key="b26304a0-f49c-11e9-8487-c5a7ea76fe8d"
        >
          <Link
            to="/completeForm?surveyId=b26304a0-f49c-11e9-8487-c5a7ea76fe8d&creatorId=MOCK_IDENTITY"
          >
            View Form
          </Link>
          <br />
          <br />
          <Link
            to="/marketplace?search=ap-southeast-2:91ecfa27-a7c0-45a8-9f45-c93885409ca7"
          >
            View Other Forms By This User
          </Link>
          <br />
          <br />
          <strong>
            Form Image
          </strong>
          <br />
          <FormImage
            image="https://designshack.net/wp-content/uploads/placeholder-image.png"
          />
        </CollapsePanel>
        <CollapsePanel
          header={
            <div
              style={
                Object {
                  "display": "flex",
                  "flexDirection": "row",
                  "justifyContent": "space-between",
                  "minWidth": 170,
                }
              }
            >
              <span
                style={
                  Object {
                    "marginTop": 4,
                  }
                }
              >
                After-Edit
              </span>
              <div>
                <Button
                  block={false}
                  ghost={false}
                  htmlType="button"
                  loading={false}
                  onClick={[Function]}
                  type="primary"
                >
                  Approve
                </Button>
                <Button
                  block={false}
                  ghost={false}
                  htmlType="button"
                  loading={false}
                  onClick={[Function]}
                  style={
                    Object {
                      "marginLeft": 4,
                    }
                  }
                  type="danger"
                >
                  Reject
                </Button>
              </div>
            </div>
          }
          key="ba665a40-f58b-11e9-9b29-99ffbf733bb5"
        >
          <Link
            to="/completeForm?surveyId=ba665a40-f58b-11e9-9b29-99ffbf733bb5&creatorId=MOCK_IDENTITY"
          >
            View Form
          </Link>
          <br />
          <br />
          <Link
            to="/marketplace?search=ap-southeast-2:91ecfa27-a7c0-45a8-9f45-c93885409ca7"
          >
            View Other Forms By This User
          </Link>
          <br />
          <br />
          <strong>
            Form Image
          </strong>
          <br />
          <FormImage
            image="https://publicus-file-uploads.s3.amazonaws.com/form/026300a0-f58c-11e9-9232-7f3bf77ba83a"
          />
        </CollapsePanel>
        <CollapsePanel
          header={
            <div
              style={
                Object {
                  "display": "flex",
                  "flexDirection": "row",
                  "justifyContent": "space-between",
                  "minWidth": 170,
                }
              }
            >
              <span
                style={
                  Object {
                    "marginTop": 4,
                  }
                }
              >
                b
              </span>
              <div>
                <Button
                  block={false}
                  ghost={false}
                  htmlType="button"
                  loading={false}
                  onClick={[Function]}
                  type="primary"
                >
                  Approve
                </Button>
                <Button
                  block={false}
                  ghost={false}
                  htmlType="button"
                  loading={false}
                  onClick={[Function]}
                  style={
                    Object {
                      "marginLeft": 4,
                    }
                  }
                  type="danger"
                >
                  Reject
                </Button>
              </div>
            </div>
          }
          key="4fe4b2c0-f4a5-11e9-87b0-6785951dc893"
        >
          <Link
            to="/completeForm?surveyId=4fe4b2c0-f4a5-11e9-87b0-6785951dc893&creatorId=MOCK_IDENTITY"
          >
            View Form
          </Link>
          <br />
          <br />
          <Link
            to="/marketplace?search=ap-southeast-2:91ecfa27-a7c0-45a8-9f45-c93885409ca7"
          >
            View Other Forms By This User
          </Link>
          <br />
          <br />
          <strong>
            Form Image
          </strong>
          <br />
          <FormImage
            image="https://designshack.net/wp-content/uploads/placeholder-image.png"
          />
        </CollapsePanel>
      </Collapse>
    </div>
  </div>
</div>
`;
